packages/ddp-server/livedata_server.js

```diff
@@ -1643,25 +1653,34 @@ _.extend(Server.prototype, {
       throw new Error("Can't call setUserId on a server initiated method call");
     };
     var connection = null;
-    var currentInvocation = DDP._CurrentInvocation.get();
-    if (currentInvocation) {
-      userId = currentInvocation.userId;
+    var currentMethodInvocation = DDP._CurrentMethodInvocation.get();
+    var currentPublicationInvocation = DDP._CurrentPublicationInvocation.get();
+    var randomSeed = null;
+    if (currentMethodInvocation) {
+      userId = currentMethodInvocation.userId;
+      setUserId = function(userId) {
+        currentMethodInvocation.setUserId(userId);
+      };
+      connection = currentMethodInvocation.connection;
+      randomSeed = DDPCommon.makeRpcSeed(currentMethodInvocation, name);
+    } else if (currentPublicationInvocation) {
+      userId = currentPublicationInvocation.userId;
       setUserId = function(userId) {
-        currentInvocation.setUserId(userId);
+        currentPublicationInvocation._session._setUserId(userId);
```

Comment:
from what I can tell there is no `setUserId` on the publication context, that's why we need to reach into `_session` to get that function. Calling `_session._setUserId` from inside a publish function should just change the current connections `userId` and the publication should re-run.